### PR TITLE
Zas atmos zone

### PR DESCRIPTION
**Pull request by TheBoondock**
## About The Pull Request
This pr adds a new proc to detect an atmos isolated area similar to detect_room however, this proc will ignore walls and non atmos_can_pass essentially an zas atmos zone. 

## Why It's Good For The Game
@LemonInTheDark has requested me to make this proc in pr #68325 but i feel like that pr is getting too big so i want to pr this separately to get some reviews and code advice

## Changelog

:cl:
code: add proc create_atmos_zone which detect an isolated atmos zone while ignoring walls and non atmos_can_pass tiles (zas zone)
/:cl:

### CODE REVIEW

**Comment by Kapu1178** [comment]
In the end... ZAS always wins...

**Comment by Colovorat** [comment]
Zased

**Review by MrMelbert — CHANGES_REQUESTED** [review]

File: code/__HELPERS/areas.dm [hunk 1]
```diff
@@ -40,6 +40,43 @@ GLOBAL_LIST_INIT(typecache_powerfailure_safe_areas, typecacheof(/area/station/en
 				continue
 			found_turfs += checkT // Since checkT is connected, add it to the list to be processed
 
+/**
+ * Create an atmos zone (Think ZAS), similiar to [proc/detect_room] but it ignores walls and turfs which are non-[atmos_can_pass]
+ *
+ * Arguments
+ * source - the turf which to find all connected atmos turfs
+ * range - the max range to check
```
Comment (LemonInTheDark) [inline code/__HELPERS/areas.dm:48]: range should be size

File: code/__HELPERS/areas.dm [hunk 2]
```diff
@@ -40,6 +40,43 @@ GLOBAL_LIST_INIT(typecache_powerfailure_safe_areas, typecacheof(/area/station/en
...
+			if(valid_turf in connected_turfs)//if the turf is already added, skip
+				loops += 1
+				continue
+			if(length(connected_turfs) >= range)
+				return 
+			if(TURFS_CAN_SHARE(reference_turf, valid_turf))
+				loops = 0 
+				connected_turfs |= valid_turf//add that to the original list				
+		if(loops >= 7)//if the loop has gone 7 consecutive times with no new turfs added, return the result. Number is arbitrary, subject to change
+			return
```
Comment (LemonInTheDark) [inline code/__HELPERS/areas.dm:77]: This check doesn't make sense

File: code/__HELPERS/areas.dm [hunk 3]
```diff
@@ -40,6 +40,43 @@ GLOBAL_LIST_INIT(typecache_powerfailure_safe_areas, typecacheof(/area/station/en
...
+	. = connected_turfs
+	while(length(connected_turfs))
+		var/list/turf/adjacent_turfs = list(
+			get_step(connected_turfs[counter], NORTH),
+			get_step(connected_turfs[counter], SOUTH),
+			get_step(connected_turfs[counter], EAST),
+			get_step(connected_turfs[counter], WEST)
+		)// get a tile in each cardinal direction at once and add that to the list
+		for(var/turf/valid_turf in adjacent_turfs)//loop through the list and check for atmos adjacency
+			var/turf/reference_turf = connected_turfs[counter]
```
Comment (LemonInTheDark) [inline code/__HELPERS/areas.dm:67]: Rather then iterating all turfs in the 4 cardinal directions, and THEN checking if they're in our atmos_adjacent turf list, why not just iterate over our atmos adjacent turfs list?

File: code/__HELPERS/areas.dm [hunk 4]
```diff
@@ -40,6 +40,43 @@ GLOBAL_LIST_INIT(typecache_powerfailure_safe_areas, typecacheof(/area/station/en
...
+			if(valid_turf in connected_turfs)//if the turf is already added, skip
+				loops += 1
+				continue
+			if(length(connected_turfs) >= range)
+				return 
+			if(TURFS_CAN_SHARE(reference_turf, valid_turf))
+				loops = 0 
+				connected_turfs |= valid_turf//add that to the original list
```
Comment (LemonInTheDark) [inline code/__HELPERS/areas.dm:75]: I don't really understand the purpose of the loops variable

File: code/__HELPERS/areas.dm [hunk 5]
```diff
@@ -40,6 +40,43 @@ GLOBAL_LIST_INIT(typecache_powerfailure_safe_areas, typecacheof(/area/station/en
...
+			get_step(connected_turfs[counter], EAST),
+			get_step(connected_turfs[counter], WEST)
+		)// get a tile in each cardinal direction at once and add that to the list
+		for(var/turf/valid_turf in adjacent_turfs)//loop through the list and check for atmos adjacency
+			var/turf/reference_turf = connected_turfs[counter]
+			if(valid_turf in connected_turfs)//if the turf is already added, skip
+				loops += 1
+				continue
+			if(length(connected_turfs) >= range)
+				return
```
Comment (LemonInTheDark) [inline code/__HELPERS/areas.dm:72]: You should do this check AFTER the add so we exit there

File: code/__HELPERS/areas.dm [hunk 6]
```diff
@@ -40,6 +40,43 @@ GLOBAL_LIST_INIT(typecache_powerfailure_safe_areas, typecacheof(/area/station/en
...
+ *
+ * Returns a list of turfs, which is an area of isolated atmos
+ */ 
+/proc/create_atmos_zone(turf/source, range = INFINITY)
+	var/counter = 1 // a counter which increment each loop
+	var/loops = 0
+	if(source.blocks_air)
+		return
+	var/list/connected_turfs = list(source)
+	. = connected_turfs
```
Comment (LemonInTheDark) [inline code/__HELPERS/areas.dm:58]: there's no need to do this, just explicitly return connected_turfs

**Review by LemonInTheDark — COMMENTED** [review]
Boon requested a review, so here ya go brother

**Pull request merged**